File: pipelines/founder_research.py
from __future__ import annotations
import asyncio
import json
from datetime import datetime
from pathlib import Path
from typing import Callable
import yaml

from providers import get_provider
from tools import RESEARCH_TOOLS, SYNTHESIS_TOOLS
from tools.web_search import web_search
from tools.read_file import read_file
from tools.browse_url import browse_url
from memory.store import recall

_PROMPTS_DIR = Path(__file__).parent.parent / "prompts"


def _load(name: str) -> dict:
    with open(_PROMPTS_DIR / f"{name}.yaml", encoding="utf-8") as f:
        return yaml.safe_load(f)


# ── Stage 1: Entity Resolution ────────────────────────────────────────────────

async def resolve_entity(input_text: str, send: Callable) -> dict:
    send({"type": "stage", "name": "entity_resolution", "status": "start", "label": "实体识别中"})

    # Check cross-session memory first
    existing = recall(input_text)
    if existing:
        send({"type": "memory_hit", "entity": existing})

    p = _load("entity_resolution")
    system = _load("system")["role"]
    task = p["task"].format(input=input_text)

    provider = get_provider()
    raw = await provider.react_loop(
        system=system,
        user=task,
        tools=[web_search],
        send=send,
        max_iterations=4,
    )

    # Extract JSON from response
    entity = _parse_json(raw)
    if not entity:
        entity = {
            "founder": input_text,
            "founder_en": "",
            "company": input_text,
            "valuation": "unknown",
            "confirmed": False,
            "confidence": "low",
        }

    send({"type": "stage", "name": "entity_resolution", "status": "done", "result": entity})
    return entity


# ── Stage 2a: Research Plan ───────────────────────────────────────────────────

async def plan_research(entity: dict, send: Callable) -> dict:
    send({"type": "stage", "name": "research_plan", "status": "start", "label": "制定搜索计划"})

    p = _load("research_plan")
    system = p["system"]
    task = p["task"].format(
        founder=entity.get("founder", ""),
        founder_en=entity.get("founder_en", "") or "",
        company=entity.get("company", ""),
        valuation=entity.get("valuation", "unknown"),
    )

    provider = get_provider()
    raw = await provider.complete(system=system, user=task)
    plan = _parse_json(raw) or {"dimensions": []}

    send({"type": "plan_ready", "plan": plan})
    send({"type": "stage", "name": "research_plan", "status": "done"})
    return plan


# ── Stage 2b: Parallel Search ─────────────────────────────────────────────────

async def execute_searches(plan: dict, send: Callable) -> dict:
    send({"type": "stage", "name": "search", "status": "start", "label": "并行搜索中"})

    async def run_query(query: str) -> dict:
        send({"type": "tool_start", "name": "web_search"})
        result = await web_search.handler(query=query)
        return {"query": query, "results": result.get("results", [])}

    tasks = []
    dimension_map = {}
    for dim in plan.get("dimensions", []):
        for q in dim.get("queries", []):
            idx = len(tasks)
            tasks.append(run_query(q))
            dimension_map[idx] = dim["name"]

    all_results = await asyncio.gather(*tasks, return_exceptions=True)

    by_dimension: dict[str, list] = {}
    for i, result in enumerate(all_results):
        dim_name = dimension_map.get(i, "misc")
        if dim_name not in by_dimension:
            by_dimension[dim_name] = []
        if not isinstance(result, Exception):
            by_dimension[dim_name].append(result)

    send({"type": "stage", "name": "search", "status": "done"})
    return by_dimension


# ── Stage 3: Synthesis ────────────────────────────────────────────────────────

async def synthesize(
    entity: dict,
    results_by_dimension: dict,
    extra_context: str,
    send: Callable,
) -> str:
    send({"type": "stage", "name": "synthesis", "status": "start", "label": "综合分析中"})

    # Format research results
    sections = []
    for dim_name, queries in results_by_dimension.items():
        sections.append(f"\n### {dim_name}\n")
        for q_result in queries:
            sections.append(f"**搜索：{q_result['query']}**")
            for r in q_result.get("results", [])[:5]:
                sections.append(f"- [{r['title']}]({r['url']})\n  {r['content'][:300]}")
    research_results = "\n".join(sections)

    p = _load("founder_profile")
    system = _load("system")["role"]
    task = p["task"].format(
        founder=entity.get("founder", ""),
        company=entity.get("company", ""),
        date=datetime.now().strftime("%Y-%m-%d"),
        valuation=entity.get("valuation", "unknown"),
        research_results=research_results,
        extra_context=f"\n### 附加上下文\n\n{extra_context}" if extra_context else "",
    )

    provider = get_provider()
    profile = await provider.react_loop(
        system=system,
        user=task,
        tools=SYNTHESIS_TOOLS,
        send=send,
        max_iterations=8,
    )

    send({"type": "stage", "name": "synthesis", "status": "done"})
    return profile


# ── Main entry point ──────────────────────────────────────────────────────────

async def run_founder_research(
    input_text: str,
    send: Callable,
    context_files: list[str] | None = None,
    context_urls: list[str] | None = None,
) -> dict:
    entity = await resolve_entity(input_text, send)

    # Pre-fetch local files and URLs before synthesis
    extra_parts = []
    for path in (context_files or []):
        send({"type": "stage", "name": "context", "status": "start", "label": f"读取 {path}"})
        result = await read_file.handler(file_path=path)
        if result.get("error"):
            send({"type": "context_error", "path": path, "error": result["error"]})
        else:
            extra_parts.append(f"**文件：{path}**\n\n{result['content']}")
        send({"type": "stage", "name": "context", "status": "done"})

    for url in (context_urls or []):
        send({"type": "stage", "name": "context", "status": "start", "label": f"获取 {url}"})
        result = await browse_url.handler(url=url)
        if result.get("error"):
            send({"type": "context_error", "path": url, "error": result["error"]})
        else:
            extra_parts.append(f"**URL：{url}**\n\n{result['content']}")
        send({"type": "stage", "name": "context", "status": "done"})

    extra_context = "\n\n---\n\n".join(extra_parts)

    plan = await plan_research(entity, send)
    results_by_dimension = await execute_searches(plan, send)
    profile = await synthesize(entity, results_by_dimension, extra_context, send)

    return {"profile": profile, "entity": entity}


# ── Helpers ───────────────────────────────────────────────────────────────────

def _parse_json(text: str) -> dict | None:
    import re
    # Strip markdown code fences
    text = re.sub(r"```(?:json)?\n?", "", text).strip().rstrip("`")
    # Find first {...}
    m = re.search(r"\{.*\}", text, re.DOTALL)
    if m:
        try:
            return json.loads(m.group())
        except json.JSONDecodeError:
            pass
    return None

File: pipelines/founder_research.js
require('dotenv').config();
const yaml = require('js-yaml');
const fs = require('fs');
const path = require('path');
const { getProvider } = require('../providers');
const { qualityCheck } = require('../eval/quality_check');
const { recall, remember } = require('../memory/store');
const { webSearch, formatResults } = require('../tools/web_search');

function loadPrompt(name) {
  const p = path.join(__dirname, '../prompts', `${name}.yaml`);
  return yaml.load(fs.readFileSync(p, 'utf8'));
}

// Stage 1: resolve arbitrary input → confirmed entity (ReAct, max 4 iterations)
async function resolveEntity(input, send) {
  send({ type: 'stage', name: 'entity_resolution', status: 'start', label: '实体识别中' });

  const prompt = loadPrompt('entity_resolution');
  const userPrompt = prompt.task.replace(/\{input\}/g, input);
  const systemPrompt = '你是一级市场研究员。通过搜索确认用户输入对应的华人创始人和公司，然后以纯 JSON 格式输出，不要有任何其他文字。';

  const { reactLoop } = getProvider();
  const raw = await reactLoop(systemPrompt, userPrompt, send, 4);

  const match = raw.match(/\{[\s\S]*?\}/);
  if (!match) throw new Error('实体识别失败：无法解析 JSON 输出');

  const entity = JSON.parse(match[0]);
  send({ type: 'stage', name: 'entity_resolution', status: 'done', result: entity });
  return entity;
}

// Stage 2a: generate a structured search plan (single LLM call, no tools)
async function planResearch(entity, send) {
  send({ type: 'stage', name: 'planning', status: 'start', label: '制定研究计划' });

  const prompt = loadPrompt('research_plan');
  const userPrompt = prompt.task
    .replace(/\{founder\}/g, entity.founder)
    .replace(/\{founder_en\}/g, entity.founder_en || entity.founder)
    .replace(/\{company\}/g, entity.company)
    .replace(/\{valuation\}/g, entity.valuation || '未知');

  const { complete } = getProvider();
  const raw = await complete(prompt.system, userPrompt);

  let plan;
  try {
    const match = raw.match(/\{[\s\S]*\}/);
    plan = JSON.parse(match ? match[0] : raw);
  } catch {
    throw new Error('研究计划生成失败：无法解析 JSON');
  }

  send({ type: 'stage', name: 'planning', status: 'done' });
  send({ type: 'plan_ready', plan });
  return plan;
}

// Stage 2b: execute all searches in parallel (no LLM, direct Tavily calls)
async function executeSearches(plan, send) {
  send({ type: 'stage', name: 'search', status: 'start', label: '并行搜索中' });

  const allQueries = plan.dimensions.flatMap(dim =>
    dim.queries.map(q => ({ query: q, dimension: dim.name, label: dim.label }))
  );

  const CONCURRENCY = 5;
  const resultsByDimension = {};
  for (const dim of plan.dimensions) resultsByDimension[dim.name] = { label: dim.label, items: [] };

  for (let i = 0; i < allQueries.length; i += CONCURRENCY) {
    const batch = allQueries.slice(i, i + CONCURRENCY);
    await Promise.all(batch.map(async ({ query, dimension }) => {
      send({ type: 'tool_start', name: 'web_search', input: { query } });
      const { results, error } = await webSearch(query);
      send({ type: 'tool_done', name: 'web_search' });
      resultsByDimension[dimension].items.push({
        query,
        content: error ? `搜索失败: ${error}` : formatResults(results),
      });
    }));
  }

  send({ type: 'stage', name: 'search', status: 'done', total: allQueries.length });
  return resultsByDimension;
}

// Format all search results into a structured string for the synthesis prompt
function formatResearchResults(resultsByDimension) {
  return Object.values(resultsByDimension).map(dim => {
    const items = dim.items.map(item =>
      `### 搜索：${item.query}\n${item.content}`
    ).join('\n\n');
    return `## ${dim.label}\n\n${items}`;
  }).join('\n\n---\n\n');
}

// Stage 3: synthesize all results into a structured profile (single LLM call, no tools)
async function synthesize(entity, resultsByDimension, extraContext, send) {
  send({ type: 'stage', name: 'synthesis', status: 'start', label: '综合分析中' });

  const sys = loadPrompt('system');
  const prof = loadPrompt('founder_profile');

  const researchResults = formatResearchResults(resultsByDimension);

  // Inject cross-session memory if prior research exists
  const memories = recall(`${entity.founder} ${entity.company}`);
  let memoryNote = '';
  if (memories.length > 0) {
    const mem = memories[0];
    const hasProfile = mem.profilePath && fs.existsSync(mem.profilePath);
    memoryNote = `\n\n> 记忆提示：此前已于 ${mem.lastResearched} 研究过该 founder。`;
    if (hasProfile) {
      memoryNote += `已有档案：\`${mem.profilePath}\`，如有需要可通过 read_file 工具读取对比。`;
    }
  }

  const userPrompt = prof.task
    .replace(/\{founder\}/g, entity.founder)
    .replace(/\{founder_en\}/g, entity.founder_en || entity.founder)
    .replace(/\{company\}/g, entity.company)
    .replace(/\{valuation\}/g, entity.valuation || '未知')
    .replace(/\{date\}/g, new Date().toLocaleDateString('zh-CN'))
    .replace(/\{research_results\}/g, researchResults)
    .replace(/\{extra_context\}/g, extraContext ? `**用户补充资料：**\n${extraContext}` : '');

  // Stage 3 uses reactLoop with browse_url only (Plan B: agent fetches full articles when needed)
  const { reactLoop, SYNTHESIS_TOOLS } = getProvider();
  const profile = await reactLoop(sys.role + memoryNote, userPrompt, send, 8, SYNTHESIS_TOOLS);

  send({ type: 'stage', name: 'synthesis', status: 'done' });
  return profile;
}

// Main entry point
// contextFiles: local file paths, contextUrls: paywalled/specific URLs to pre-fetch
async function runFounderResearch(input, send, contextFiles = [], contextUrls = []) {
  // Stage 1
  const entity = await resolveEntity(input, send);
  if (!entity.confirmed) {
    throw new Error(`无法确认「${input}」对应估值 2 亿美金以上的华人创始人，请提供更多信息`);
  }

  // Load user-provided local files and URLs
  let extraContext = '';
  const hasContext = contextFiles.length > 0 || contextUrls.length > 0;
  if (hasContext) {
    const { readFile } = require('../tools/read_file');
    const { browseUrl } = require('../tools/browse_url');
    send({ type: 'stage', name: 'context', status: 'start', label: '读取补充资料' });

    for (const filePath of contextFiles) {
      const { content, error } = await readFile(filePath);
      if (content) extraContext += `\n\n[文件: ${path.basename(filePath)}]\n${content}`;
      else send({ type: 'context_error', path: filePath, error });
    }

    for (const url of contextUrls) {
      const { content, error } = await browseUrl(url);
      if (content) extraContext += `\n\n[网页: ${url}]\n${content}`;
      else send({ type: 'context_error', path: url, error });
    }

    send({ type: 'stage', name: 'context', status: 'done' });
  }

  // Stage 2a: Plan
  const plan = await planResearch(entity, send);

  // Stage 2b: Execute
  const resultsByDimension = await executeSearches(plan, send);

  // Stage 3: Synthesize
  const profile = await synthesize(entity, resultsByDimension, extraContext, send);

  const metrics = qualityCheck(profile, entity);
  send({ type: 'eval', metrics });

  return { profile, entity };
}

module.exports = { runFounderResearch, remember };
